File: python/mirage/commands/builtin/generic/comm.py
from collections.abc import Awaitable, Callable

from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


def _comm_merge(lines1: list[str], lines2: list[str]) -> list[tuple[int, str]]:
    result: list[tuple[int, str]] = []
    i, j = 0, 0
    while i < len(lines1) and j < len(lines2):
        if lines1[i] < lines2[j]:
            result.append((1, lines1[i]))
            i += 1
        elif lines1[i] > lines2[j]:
            result.append((2, lines2[j]))
            j += 1
        else:
            result.append((3, lines1[i]))
            i += 1
            j += 1
    while i < len(lines1):
        result.append((1, lines1[i]))
        i += 1
    while j < len(lines2):
        result.append((2, lines2[j]))
        j += 1
    return result


def _format_comm(
    merged: list[tuple[int, str]],
    suppress1: bool,
    suppress2: bool,
    suppress3: bool,
) -> str:
    out: list[str] = []
    for col, text in merged:
        if col == 1 and not suppress1:
            out.append(text)
        elif col == 2 and not suppress2:
            prefix = "" if suppress1 else "\t"
            out.append(prefix + text)
        elif col == 3 and not suppress3:
            prefix = ""
            if not suppress1:
                prefix += "\t"
            if not suppress2:
                prefix += "\t"
            out.append(prefix + text)
    return "\n".join(out) + "\n" if out else ""


async def comm(
    paths: list[PathSpec],
    *,
    read_bytes: Callable[..., Awaitable[bytes]],
    accessor: object = None,
    suppress1: bool = False,
    suppress2: bool = False,
    suppress3: bool = False,
    check_order: bool = False,
) -> tuple[ByteSource | None, IOResult]:
    if len(paths) < 2:
        raise ValueError("comm: requires two paths")
    data1 = (await read_bytes(accessor, paths[0])).decode(errors="replace")
    data2 = (await read_bytes(accessor, paths[1])).decode(errors="replace")
    lines1 = data1.splitlines()
    lines2 = data2.splitlines()
    stderr = ""
    if check_order:
        if lines1 != sorted(lines1):
            stderr = "comm: file 1 is not in sorted order\n"
        elif lines2 != sorted(lines2):
            stderr = "comm: file 2 is not in sorted order\n"
    merged = _comm_merge(lines1, lines2)
    output = _format_comm(merged, suppress1, suppress2, suppress3)
    return output.encode(), IOResult(
        stderr=stderr.encode() if stderr else None)


__all__ = ["comm"]

File: typescript/packages/core/src/commands/builtin/generic/comm.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { IOResult, materialize, type ByteSource } from '../../../io/types.ts'
import type { PathSpec } from '../../../types.ts'
import type { CommandFnResult, CommandOpts } from '../../config.ts'

const ENC = new TextEncoder()
const DEC = new TextDecoder('utf-8', { fatal: false })

function splitLinesNoTrailing(text: string): string[] {
  const stripped = text.endsWith('\n') ? text.slice(0, -1) : text
  return stripped === '' ? [] : stripped.split('\n')
}

type MergeEntry = [number, string]

function commMerge(lines1: readonly string[], lines2: readonly string[]): MergeEntry[] {
  const result: MergeEntry[] = []
  let i = 0
  let j = 0
  while (i < lines1.length && j < lines2.length) {
    const a = lines1[i] ?? ''
    const b = lines2[j] ?? ''
    if (a < b) {
      result.push([1, a])
      i += 1
    } else if (a > b) {
      result.push([2, b])
      j += 1
    } else {
      result.push([3, a])
      i += 1
      j += 1
    }
  }
  while (i < lines1.length) {
    result.push([1, lines1[i] ?? ''])
    i += 1
  }
  while (j < lines2.length) {
    result.push([2, lines2[j] ?? ''])
    j += 1
  }
  return result
}

function formatComm(
  merged: readonly MergeEntry[],
  suppress1: boolean,
  suppress2: boolean,
  suppress3: boolean,
): string {
  const out: string[] = []
  for (const [col, text] of merged) {
    if (col === 1 && !suppress1) {
      out.push(text)
    } else if (col === 2 && !suppress2) {
      const prefix = suppress1 ? '' : '\t'
      out.push(prefix + text)
    } else if (col === 3 && !suppress3) {
      let prefix = ''
      if (!suppress1) prefix += '\t'
      if (!suppress2) prefix += '\t'
      out.push(prefix + text)
    }
  }
  return out.length > 0 ? out.join('\n') + '\n' : ''
}

function isSorted(lines: readonly string[]): boolean {
  for (let i = 1; i < lines.length; i++) {
    if ((lines[i - 1] ?? '') > (lines[i] ?? '')) return false
  }
  return true
}

export async function commGeneric(
  paths: PathSpec[],
  opts: CommandOpts,
  stream: (p: PathSpec) => AsyncIterable<Uint8Array>,
): Promise<CommandFnResult> {
  if (paths.length < 2) {
    return [null, new IOResult({ exitCode: 1, stderr: ENC.encode('comm: requires two paths\n') })]
  }
  const p1 = paths[0]
  const p2 = paths[1]
  if (p1 === undefined || p2 === undefined) return [null, new IOResult()]
  const data1 = DEC.decode(await materialize(stream(p1)))
  const data2 = DEC.decode(await materialize(stream(p2)))
  const lines1 = splitLinesNoTrailing(data1)
  const lines2 = splitLinesNoTrailing(data2)
  let stderr = ''
  if (opts.flags['check-order'] === true) {
    if (!isSorted(lines1)) stderr = 'comm: file 1 is not in sorted order\n'
    else if (!isSorted(lines2)) stderr = 'comm: file 2 is not in sorted order\n'
  }
  const suppress1 = opts.flags.args_1 === true
  const suppress2 = opts.flags['2'] === true
  const suppress3 = opts.flags['3'] === true
  const merged = commMerge(lines1, lines2)
  const output = formatComm(merged, suppress1, suppress2, suppress3)
  const result: ByteSource = ENC.encode(output)
  return [result, new IOResult({ stderr: stderr !== '' ? ENC.encode(stderr) : null })]
}
